### ISSUE TITLE
Error: C stack usage is too close to the limit

### ISSUE THREAD

**Issue by altairwei** [StatBiomed/UniTVelo#22]
Hello,

I recently upgraded UniTVelo from v0.2.3 to v0.2.5 and found that it didn't work. UniTVelo v0.2.3 was successfully run with the same data and codes.

The following is the log:

```
(Running UniTVelo 0.2.5)
2022-12-08 15:16:21
------> Manully Specified Parameters <------
NUM_REP:        2
NUM_REP_TIME:   re_init
BASIS:  pca
------> Model Configuration Settings <------
N_TOP_GENES:    2000
LEARNING_RATE:  0.01
FIT_OPTION:     1
DENSITY:        SVD
REORDER_CELL:   Soft_Reorder
AGGREGATE_T:    True
R2_ADJUST:      True
GENE_PRIOR:     None
VGENES: basic
IROOT:  None
--------------------------------------------

Current working dir is /home/XXX.
Results will be stored in res folder
Filtered out 54980 genes that are detected 20 counts (shared).
WARNING: Did not normalize X as it looks processed already. To enforce normalization, set `enforce=True`.
Normalized count data: spliced, unspliced.
Extracted 2000 highly variable genes.
WARNING: Did not modify X as it looks preprocessed already.
Extracted 2000 highly variable genes.
Computing moments for 2000 genes with n_neighbors: 30 and n_pcs: 30
computing neighbors
    finished (0:00:08) --> added
    'distances' and 'connectivities', weighted adjacency matrices (adata.obsp)
computing moments based on connectivities
    finished (0:00:01) --> added
    'Ms' and 'Mu', moments of un/spliced abundances (adata.layers)

# of velocity genes 982 (Criterion: positive regression coefficient between un/spliced counts)
# of velocity genes 953 (Criterion: std of un/spliced reads should be moderate, w/o extreme values)
# of velocity genes 949 (Criterion: genes have reads in more than 5% of total cells)
Using GPU card: 0
Loss (Total): 58018.953, (Spliced): 30779.621, (Unspliced): 27239.332:  38%|██████████████████▌                              | 4545/12000 [11:18<15:35,  7.97it/s]Build failed. See log files for full details.
Error: C stack usage  7970020 is too close to the limit
```

**Comment by michaelgmz** [comment]
Hi altairwei, 

Sorry to hear this.
I tested this version with a few hands on datasets with cells ranging from 3k to 30k, unfortunately I haven't found the above error.
It seems weird but I suspect this might related to some external memory problem other than this package, I'll try investigating this further.

Does the optimization procedure always been interrupted at this particular iteration 4545 or at random steps?
Could you please try creating a new conda environment and see if the package works?

Sorry for the inconvenience.

Bests,
Mingze

**Comment by altairwei** [comment]
Hi @michaelgmz,

Does the optimization procedure allocate memory on VRAM? I monitored RAM during the whole process, and didn't find any signs of insufficient memory. However, I forgot to monitor the VRAM usage, maybe it has something to do with that? I will try again to see if I can reproduce the problem.

Bests,
Altair

**Comment by altairwei** [comment]
Hi @michaelgmz 

After digging into the UniTVelo source code and a few trials, I found the root of the problem. The error was caused by the R package `knitr`, and yes I was calling UniTVelo through `reticulate`. My guess is that when rendering RMarkdown documents, `knitr` stores UniTVelo's output text, including the progress bar text, and that the progress bar text depletes R's stack memory after about 5000 cycles. I didn't find any problem calling UniTVelo directly from the R terminal using `reticulate`.

Bests,
Altair
